src/script/interpreter.cpp

```diff
@@ -1962,6 +2023,60 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 }
                 break;
 
+                case OP_ECMULSCALARVERIFY:
+                {
+                    // OP_ECMULSCALARVERIFY is available post tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+
+                    valtype& vchRes = stacktop(-3);
+                    valtype& vchGenerator = stacktop(-2);
+                    valtype& vchScalar = stacktop(-1);
+
+                    CPubKey pk(vchGenerator);
+                    CPubKey res(vchRes);
+                    if (!pk.IsFullyValid() || !res.IsFullyValid() || !pk.IsCompressed() || !res.IsCompressed())
+                        return set_error(serror, SCRIPT_ERR_PUBKEYTYPE);
+                    execdata.m_validation_weight_left -= VALIDATION_WEIGHT_PER_SIGOP_PASSED;
+                    if (execdata.m_validation_weight_left < 0) {
+                        return set_error(serror, SCRIPT_ERR_TAPSCRIPT_VALIDATION_WEIGHT);
+                    }
+                    if (vchScalar.size() != 32 || !pk.TweakMulVerify(res, uint256(vchScalar)))
+                        return set_error(serror, SCRIPT_ERR_ECMULTVERIFYFAIL);
+
+                    popstack(stack);
+                    popstack(stack);
+                    popstack(stack);
+                }
+                break;
+                case OP_TWEAKVERIFY:
+                {
+                    // OP_TWEAKVERIFY is available post tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+
+                    valtype& vchTweakedKey = stacktop(-3);
+                    valtype& vchTweak = stacktop(-2);
+                    valtype& vchInternalKey = stacktop(-1);
+
+                    CPubKey tweakedKey(vchTweakedKey);
+                    if (!tweakedKey.IsFullyValid() || !tweakedKey.IsCompressed() || vchInternalKey.size() != 32 || vchTweak.size() != 32)
```

Comment:
Checking FullyValid is an unnecessary expense.  I believe IsCompressed is good enough.

Reply:
I did consider this, but there is still some utility in that this returns a different error type.

Reply:
I can also just substitute it for the simple check for the first byte 0x02

Reply:
This is more efficient, part of me was over checking things because of the aborts :)

Reply:
IsCompressed is fine.  It implies 0x02 and 0x03.  My concern with FullyValid is that it involves the unprecedented and moderately expensive crypto operation of sqrt.

Reply:
I directly checked the first byte instead of allocating a new `CPubKey` structure which was not required.